README.md
# xbee-serialSim

Simulation of serial line communication for xbee-api framework

Development with the xbee-api (https://github.com/jankolkmeier/xbee-api) normally need usage of running physical xbee modules.

I was working on a small project to realize my own smart home system and have used xbee-api to catch the ZigBee frames from serial line, work with them and push commands back into my network.

To be more flexible i developed a simulation for the serialport module for nodejs, created by voodootikigod: https://github.com/voodootikigod/node-serialport


## configuration

The nodes will be represented by an arry of objects:

    var nodeList = [

        {
            shortMac: "01A1",
            longMac: "0013A2000000001",
            LckInterval: 20,
            DateInterval: 30,
            data: "AT+1=23.15",
            type: "Temperature Sensor"

        },
        {
            shortMac: "01A2",
            longMac: "0013A2000000002",
            LckInterval: 30,
            DateInterval: 30,
            data: "AT+1=0",
            type: "Door Contact Sensor"

        }

    ];

* ShortMac: The 16 bit short mac address
* LongMac: the 64 bit long mac address
* LckInterval: The interval in which the node shall send a lifecheck command: AT+1=OK
* DateInterval: The interval in which the node shall send the at data saved command string
* data: Command string, encoded in AT syntax
* type: Not used at the moment just to clarify what sort of sensor is simulated

This is the first version which should provide basic functionality. During my development (which is a private project) i will add more features in future releases.

## Limits of the system

To keep it simple, the simulation don´t emit raw buffer frames but JSON objects according to the syntax and structure which is used in the xbee-api.

## Using the simulation

At the test.js a simple usage is shown.

The basic usage starts at line 70 where the serialsim.open() function is called. There is either a callback or an event 'open' emitted after success.

After that, the nodeSim.start() function should be called to start the simulation.

For each node, two events will be registered, which emmits an event after the defined interval. one for the lifecheck and one for the dateinterval.

These events creates JSON objects and emmits them outside to be catched by the serialSim object. This ojbect catches the event and emmits directly in the xbee-api object the event which passes the JSON object.

Writing to the serial line simulation works similar:

The JSON object will be passed to the serialSim object function write() which emmits an event to be catched by a function. This function passes the JSON object to the nodeSim function receiveFrame() which will parse the object and reacts according to frame type, commands, values etc.


## Adjusting the code:

There are two locations within the code where the simulation should be adjusted:

1. nodes.js: Here all the information about the nodes should be changed
2. nodeSim.receiveFrame function: Here the received frames will be analysed and responses or actions will be triggered.










CHANGES.md
# Changes

## 0.0.12 (29.10.15)
* added CHANGES.md and transfered gitlog into file
* added .gitignore
* renamed folder test to spec
* added jasmine.json
* added reporterConig.js
* added module jasmine-reporters@2.0.7

## 0.0.11 (15.3.15)
* added correct object support at example nodelist at nodeSim.spec.js
* added xbee-api support for correct test pass at serialSim.spec.js
* added rootpath module;
* added support for config file "DebugMessages enabling/disabling at serialSim.js
* added object with sensor objects. at lmSHObjects.js
* added option to only print to console when debugging is enabled at nodeSim.js

## 0.0.10 (15.3.15)
* Temperature and Door contact now transmit with the correct object id. at node.js

## 0.0.9(19.2.15)
* added command to copy pack into local_modules folder (Gruntfile.js)
* prepared to react on sended at commands at nodeSim.js and serialSim.js

## 0.0.8 (29.1.15)
* fixed wrong main file at package.json
* longMac shall be 16bytes long otherwise xbee-api throws an error. at node.js
* added grunt support


## 0.0.7 (28.1.15)
* differentiate between livecheck interval and dateinterval; changed interval to more convenient length at nodes.js
* added more test cases at nodeSim.spec.js.
* nodeSim.js:
    * Added functions getNodeForLongMac and receiveFrame.
    * Added callback option for start and stop function.
    * RaiseEvent function now correctly raises frame objects.
    * Start now add different timers for livecheck and date interval
* serialSim:
    * added test cases.
    * added code description.
    * read and write function now working.
* test now performs a complete simulation for 1.5 min at test.js
* renamed project from serialSim to xbee-serialSim for clarification
* adapted test for event counter to test lifecheck and dateinterval event for each node. at nodeSim.spec.js
* created LICENCE.md
* created README.md
* created initial description at README.md
* rename project to be compliant with npm (only lower-case names are allowed).
* corrected output of getStringForDate - added seconds to string. at serialSim.js ad test.js
* nodesim is now completely encapsulated within serialSim for more simple usage. at serialSim.js



## 0.0.6 (12.1.15)
* emitting started / stopped event now after synchron for loop not inside to avoid emitting twice or too early. at nodeSim.js
* added afterEach for Start / Stop test to stop events. Added test for check EventCounter at nodeSim.spec.js
* stop() now clears the event list at nodeSim.js
* added test if after called stop() the eventlist is empty at nodeSim.spec.js

## 0.0.5 (9.1.15)
* nodeList now has to be passed as parameter at constructor; necessary to be more flexible e.g. for testing
* added function getNodelist and related tests. Added documentation for nodeSim
* nodeSim.raiseEvent now raises the complete node object as parameter during emitting event.


## 0.0.4 (9.1.15)
* initial commit of karma.conf.js and run_test
* moved nodeSim.js to folder src
* added new function EventCounter to return actual number of listed events
* added events to be thrown for functions start and stop
* moved nodes.js to folder src
* moved serialSim.js to forder src
* added dev dependencies for karma and node (testing) at package.json
* testing of emitted events at test.js

## 0.0.3 (8.1.15)
* initial commit of node simulation; start stop and raiseEvent implemented at nodeSim.js
* test for nodeSim implemented, comment previous tests at test.js

## 0.0.2 (8.1.15)
* initial test script at test.js
* exports with capital Letters to fulfill convention at serialSim.js
* initial commit of test nodes ad node.js

## 0.0.1 (8.1.15)
* initial commit; created class serialSim

test.js
/**
 * Created by lynmatten on 08.01.15.
 */


var fs = require('fs');

var xbee_api = require('xbee-api');
var C = xbee_api.constants;
var xbeeAPI = new xbee_api.XBeeAPI();


var SerialSim = require('./serialSim.js');
var serialSim = new SerialSim.SerialSim();


var events = require('events');
var eventEmitter = new events.EventEmitter();

var NodeSim = require('./nodeSim.js');
var nodeSim = new NodeSim.nodeSim();


// Something we might receive from an XBee...
var raw_frame = new Buffer([
    0x7E, 0x00, 0x13, 0x97, 0x55, 0x00, 0x13, 0xA2, 0x00, 0x40, 0x52, 0x2B,
    0xAA, 0x7D, 0x84, 0x53, 0x4C, 0x00, 0x40, 0x52, 0x2B, 0xAA, 0xF0
]);

xbeeAPI.on('data', function(data) {


    console.log(">> data received: " + data);
});

// All frames parsed by the XBee will be emitted here
xbeeAPI.on("frame_object", function(frame) {
    console.log(">> frame received: ", frame);

    var frameObj = xbeeAPI.parseFrame(frame);

    console.log('>> frame parsed: ', frameObj);
});

var bla = function bla(data) {

    console.log("bla: " + data);
    xbeeAPI.emit('data',data);
    xbeeAPI.emit('frame_object', data);
};

eventEmitter.on('data', bla);


/**
 * check OS
 */

if(process.platform != "darwin")
{
    console.log("Your operating system is not supported. Sorry!");
    process.exit(1);
}


/**
 * check if serial port exists if not create
 * **/


if(fs.existsSync('/dev/tty.zigbeeVirtual1'))
{
    /**
     * Exists, proceed
     */
}
else
{
    /**
     * Don´t exists, create fifo
     */
}


serialSim.open();



/*
eventEmitter.emit('data', raw_frame);

serialSim.write("Test Frame", function(err) {

    console.log("with callback!");
    console.log(err);

});
*/


nodeSim.start();

setTimeout(function() {
    nodeSim.stop();
    serialSim.close();

}, 30000);



/**************************************************/

/*

 var serialport = new SerialPort("/dev/tty.zbtest1", {
 baudrate: 57600

 });

 serialport.on("open", function() {

 console.log("Serial Port opened.");

 });

 */


src/test.js
/**
 * Created by lynmatten on 08.01.15.
 */


var xbee_api = require('xbee-api');
var C = xbee_api.constants;
var xbeeAPI = new xbee_api.XBeeAPI();

var nodeList = require('./nodes.js');
var SerialSim = require('./serialSim.js');
var serialSim = new SerialSim.SerialSim(xbeeAPI, nodeList);


/** Example usage **/

/******************************************/

serialSim.open(function() {

    console.log("Event: xbee-serialsim started.");



    setTimeout(function() {

        var frame_obj = { // AT Request to be sent to
            type: C.FRAME_TYPE.REMOTE_AT_COMMAND_REQUEST,
            id: 0x01,
            //destinantion64:"0013A20040ADD0D3",
            //destination64:"0013A200FFFFFFFF",
            //destination16: "0x00",
            remoteCommandOptions: 0x02,
            command: "MY",
            commandParameter: [0x01]
        };



        serialSim.write(xbeeAPI.buildFrame(frame_obj));

    }, 500);

    setTimeout(function() {

        serialSim.close();


    }, 90000);


});




/**************************************************/


Gruntfile.js
/**
 * Created by lynmatten on 30.01.15.
 */


module.exports = function(grunt) {

    var pkg = grunt.file.readJSON('package.json');

    grunt.initConfig({

        pkg: pkg,

        exec: {
            echo_grunt_version: {
                cmd: function () {
                    return 'echo ' + this.version;
                }
            },
            npm_pack: {
                cmd: 'npm pack'
            },
            delete_all_packs: {
                cmd: 'rm -rf xbee-serialsim*.tgz'
            },
            rename_actual_pack_to_latest: {
                cmd: 'cp ' + pkg.name + '-' + pkg.version + '.tgz ' + pkg.name + '.tgz'
            }
        }


    });

    grunt.registerTask('default', 'print-default-message', function() {

        grunt.log.writeln("No default task defined.");

    });

    grunt.registerTask('pack', ['exec:npm_pack','exec:rename_actual_pack_to_latest']);

    grunt.registerTask('version', 'displays name and version', function() {

        grunt.log.writeln('Name: ' + pkg.name + " -- Version: " + pkg.version);

    });


    grunt.loadNpmTasks('grunt-contrib-copy');
    grunt.loadNpmTasks('grunt-exec');

};

karma.conf.js
/**
 * Created by lynmatten on 09.01.15.
 */


// Karma configuration for unit tests

module.exports = function (config) {
    config.set({

        // base path, that will be used to resolve files and exclude
        basePath: '',


        // frameworks to use
        frameworks: ['jasmine'],


        // list of files / patterns to load in the browser
        files: [
            'src/*.js',
            'test/unit/*.js'
        ],


        // list of files to exclude
        exclude: [

        ],


        // test results reporter to use
        // possible values: 'dots', 'progress', 'junit', 'growl', 'coverage'
        reporters: ['progress'],


        // web server port
        port: 9876,


        // enable / disable colors in the output (reporters and logs)
        colors: true,


        // level of logging
        // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        //logLevel: config.LOG_DEBUG,
        logLevel: config.LOG_INFO,

        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: false,


        // Start these browsers, currently available:
        // - Chrome
        // - ChromeCanary
        // - Firefox
        // - Opera
        // - Safari (only Mac)
        // - PhantomJS
        // - IE (only Windows)
        browsers: ['Chrome'],


        // If browser does not capture in given timeout [ms], kill it
        captureTimeout: 60000,


        // Continuous Integration mode
        // if true, it capture browsers, run tests and exit
        singleRun: true
    });
};


src/serialSim.js
/**
 * Created by lynmatten on 08.01.15.
 */

require('rootpath')();

var NodeSim = require('./nodeSim.js');

var zbh = require('xbee-helper');
var helper = new zbh.ZigBeeHelper();

var events = require('events');

var genConfig = null;

try {
    genConfig = require('config.js');
}
catch(err) {
    genConfig = {};
    genConfig.DebugMessages = true;
    genConfig.DebugMessagesSimulation = true;

}

/***
 * Class xbee-serialsim
 * simulates a serial line communication for xbee-api
 * @param path
 * @param xbeeObj
 * @constructor
 */
var serialSim = function(xbeeObj, nodeList) {

    var self = this;

    events.EventEmitter.call(this);
    this._open = false;
    this._nodeList = nodeList;
    this._xbeeObj = xbeeObj;

    if(xbeeObj == undefined || xbeeObj == null)
    {
        throw new Error("No instance of xbee-api found.");
    }

    this._nodeSim = new NodeSim.NodeSim(this._nodeList);


    /**********/

    this._xbeeObj.on('data', function(data) {
        //console.log(">> data received: ", data);
    });

// All frames parsed by the XBee will be emitted here
    this._xbeeObj.on("frame_object", function(frame) {
        //console.log(">> frame received by xbeeAPI: ", frame);

        if(genConfig.DebugMessages == true && genConfig.DebugMessagesSimulation == true)
        {
            console.log(helper.printFrame(frame));
        }


    });

    var emitXbee = function (data) {

        this._xbeeObj.emit('data',data);
        this._xbeeObj.emit('frame_object', data);
    };


    var displayStart = function() {

        if(genConfig.DebugMessages == true) {


            console.log("Event: nodeSim started. -- " + self._nodeSim.EventCounter() + " Events.");
        }
    };

    var displayOpen = function() {

        console.log(" xbee-serialsim started.");
    };


    this._nodeSim.on('sendFrame', function(rawFrame) {

        self.read(rawFrame);

    });

    var passFrameToNodeSim = function(data) {

        self._nodeSim.receiveFrame(data);
    };

    this._nodeSim.on('started', displayStart);
    this.on('open', displayOpen);

    this.on('frame_object', emitXbee);

    this.on('receiveFrame', passFrameToNodeSim);

    /*********/


};

/**
 * prototype for eventemitter
 * @type {Object|Function|exports.EventEmitter}
 * @private
 */
serialSim.prototype.__proto__ = events.EventEmitter.prototype;


/**
 * function open(), synchron or asynchron
 * Emitts 'open' event at the end
 * @param callback
 */
serialSim.prototype.open = function(callback) {

    var self = this;

    //console.log("SerialSim opened.");

    this._nodeSim.start(function() {

        if(genConfig.DebugMessages == true)
        {
            console.log("Event: nodeSim started. -- " + self._nodeSim.EventCounter() + " Events.");
        }


    });

    self.emit('open');

    if(callback) { callback(); }

};

/**
 * function close(), synchron or asynchron
 * Emitts 'close' event at the end
 * @param callback
 */
serialSim.prototype.close = function(callback) {

    var self = this;

    //console.log("SerialSim closed.");

    this._nodeSim.stop(function() {
        self.emit('close');
        self.removeAllListeners();

        if(callback) { callback(); }
    });



};

/**
 * function write(), synchron or asynchron
 * simulates writing to the serial line
 * @param data
 * @param callback
 */
serialSim.prototype.write = function(data, callback) {

    var self = this;

    /* TODO: simulate writing to serial Line
      *  receive a buffer, need to xbeeAPI.parseFrame to get JSON object
      *  check with addressed node for reaction of command
       * */

    /* write */
    //console.log("writing data: ", data);
    self.emit('receiveFrame', data);

    if(callback) {

        //var err = new Error('Unknown Error while sending data: ' + data);

        callback();
    }

};

/**
 * function read(), synchron
 * simulates reading event from serial line
 * Emits 'receivedData' event with data from node
 * @param rawFrame
 */
serialSim.prototype.read = function(rawFrame) {

    this.emit('frame_object', rawFrame);

};


exports.SerialSim = serialSim;

src/nodeSim.js
/**
 * Created by lynmatten on 08.01.15.
 */

//var nodeList = require('./nodes.js');


var events = require('events');

var xbee_api = require('xbee-api');
var C = xbee_api.constants;
var xbeeAPI = new xbee_api.XBeeAPI();

var xbeeHelper = require('xbee-helper');
var ZigBeeHelper = new xbeeHelper.ZigBeeHelper();

var genConfig = null;

try {
    genConfig = require('config.js');
}
catch(err) {
    genConfig = {};
    genConfig.DebugMessages = true;
    genConfig.DebugMessagesSimulation = true;

}
var shObj = require('./lmSHObjects.js');
/**
 * Class NodeSim
 * Simulation of nodes of a Xbee network
 * @param nodeList
 * @constructor
 */
var NodeSim = function(nodeList)
{

    events.EventEmitter.call(this);

    this._nodeList = nodeList;
    this._eventList = [];
    this._timerChangeList = [];

    if(this._nodeList == null || this._nodeList === 'undefined')
    {
        throw new Error('no nodelist found.');
    }


};

/**
 *
 * @type {Object|Function|exports.EventEmitter}
 * @private
 */
NodeSim.prototype.__proto__ = events.EventEmitter.prototype;


/**
 * Starts simulation of network traffic; initialize node specific events
 */
NodeSim.prototype.start = function(callback) {

    var self = this;

    for(index = 0; index < this._nodeList.length; index++)
    {
        var timer1 = setInterval(self.raiseEvent, this._nodeList[index].LckInterval * 1000, this, this._nodeList[index].shortMac, "L");
        var timer2 = setInterval(self.raiseEvent, this._nodeList[index].DateInterval * 1000, this, this._nodeList[index].shortMac,"D");

        this._eventList.push(timer1);
        this._eventList.push(timer2);

    }
    self.emit('started');

    if(callback) { callback(); }



};

/**
 * stops simulation of netowrk traffic; removes all node specific events
 */
NodeSim.prototype.stop = function(callback) {

    var self = this;

    if(this._eventList.length == 0)
    {
        self.emit('stopped');

        if(callback) { callback(); }
        return;
    }

    for(index = 0; index < this._eventList.length; index++)
    {
        clearInterval(this._eventList[index]);

    }

    /* clear eventlist */
    this._eventList = [];
    self.emit('stopped');

    if(callback) { callback(); }


};

/**
 * emits event for node by delivereing its short mac address.
 * The parameter self is necessary because the std 'this' is in this function the object of the timer, not of the class
 * @param self
 * @param shortMac
 */
NodeSim.prototype.raiseEvent = function(self, shortMac, type) {

    var node = null;

    for (var index = 0; index < self._nodeList.length; index++)
    {
        if(self._nodeList[index].shortMac === shortMac)
        {
            node = self._nodeList[index];
        }
    }

    if(node !== null)
    {

        /****
         *
         * checking if changes waiting for node
         */
        for (var i = 0; i < self._timerChangeList.length; i++) {

            if(self._timerChangeList[i].shortMac === node.shortMac) {

                if(self._timerChangeList[i].type === type) {

                    delete self._eventList[self._eventList.indexOf(this)];
                    clearInterval(this);
                    var timer1 = setInterval(self.raiseEvent, node.LckInterval * 1000, self, node.shortMac, "L");
                    self._eventList.push(timer1);

                    console.log("# xbee-serialsim # Changed LckInterval for " + node.shortMac + " to " + node.LckInterval, "NODE");
                }
            }

        }

        /*****************/

        var dataStr = "";

        if(type == 'L')
        {
            dataStr = "AT+1=OK"
        }
        else if(type == 'D')
        {
            dataStr = node.data;
        }

        var frame_obj = {
            type: C.FRAME_TYPE.ZIGBEE_RECEIVE_PACKET,
            remote16: node.shortMac,
            remote64: node.longMac,
            receiveOptions: 0x01,
            data: ZigBeeHelper.StringToByteArray(dataStr)


        };

/*
        var packet  = new Buffer([0x7E, 0x00, 0x0F, 0x90, 0x00, 0x01,
            0x3A, 0x20, 0x00, 0x00, 0x00, 0x01, 0x01, 0xA1, 0x01,
            0xA1, 0x23, 0x15, 0x97]);

        var packetEscaped = xbeeAPI.escape(packet);

*/
        self.emit('sendFrame', frame_obj);

    }
    else
    {
        throw new Error('Node for shortMac not found.');
    }


};

/**
 * Returns the number of registered events
 * @returns {Number}
 * @constructor
 */
NodeSim.prototype.EventCounter = function() {

    return this._eventList.length;

};

/**
 * returns the complete list of simualted nodes.
 * @returns {*}
 */
NodeSim.prototype.getNodelist = function() {

    return this._nodeList;

};

/**
 * returns node object for long mac address
 * @param longMac
 * @returns {*}
 */
NodeSim.prototype.getNodeForLongMac = function(longMac)
{
    for(var i = 0; i < this._nodeList.length; i++)
    {
        if(this._nodeList[i].longMac == longMac.toUpperCase())
        {
            return this._nodeList[i];
        }
    }

    return null;
}

/**
 * simulation of receiption of xbeee raw frames
 * @param data
 */
NodeSim.prototype.receiveFrame = function(data) {

    var self = this;

    //console.log("received frame by xbee-serialsim: ", data);

    var data_obj = xbeeAPI.parseFrame(data);

    //console.log('parsed frame by xbee-serialsim:', data_obj);

    if(genConfig.DebugMessages == true && genConfig.DebugMessagesSimulation == true) {

        console.log(ZigBeeHelper.printFrame(data_obj));
    }
    var node = this.getNodeForLongMac(data_obj.destination64);

    if(node === undefined || node === null && data_obj.destination64.toUpperCase() !== '000000000000FFFF') /** if broadcast **/
    {
        throw new Error('Node for longmac not existing');
        return null;
    }



    switch(data_obj.type)
    {
        case C.FRAME_TYPE.REMOTE_AT_COMMAND_REQUEST:


            switch(data_obj.command)
            {
                case 'MY':


                    //console.log("REMOTE_AT_COMMAND_REQUEST");

                    for(var i = 0; i < this._nodeList.length; i++) {

                        var ret_obj = { type: C.FRAME_TYPE.REMOTE_COMMAND_RESPONSE,
                            //id: 1,
                            remote64: this._nodeList[i].longMac,
                            remote16: this._nodeList[i].shortMac,
                            command: 'MY',
                            commandStatus: 1,
                            commandData: []
                        };


                        var emmiterForObject = function(ret_obj_inner) {

                            self.emit('sendFrame', ret_obj_inner);
                        };

                        setTimeout(emmiterForObject, 1000 * (i+1), ret_obj);

                    }

                    /** return to avoid double emitting **/
                    return;

                    break;
            }



            break;


        case C.FRAME_TYPE.ZIGBEE_RECEIVE_PACKET:

            var cmd_obj = ZigBeeHelper.getATCommand(ZigBeeHelper.ByteToString( data_obj.data, true));

            console.log("cmd_obj: ", cmd_obj);


            break;

        case C.FRAME_TYPE.ZIGBEE_TRANSMIT_REQUEST:

            var cmd_obj = ZigBeeHelper.getATCommand(ZigBeeHelper.ByteToString( data_obj.data, true));

            //console.log("cmd_obj: ", cmd_obj);

            for (var i = 0; i < cmd_obj.commandArr.length; i++) {

                var cmdArr = cmd_obj.commandArr[i];

                //console.log(cmdArr);

                switch (parseInt(cmdArr.commandName)) {

                    case shObj.OBJ_LIFECHECK_INTERVAL:

                        console.log("# xbee-serialsim # change lckInterval for "+ node.shortMac +" to " + cmdArr.commandParam, "NODE");

                        node.LckInterval = parseInt(cmdArr.commandParam);

                        var new_Obj = {

                            'shortMac': node.shortMac,
                            'type': 'L',
                            'newVal': parseInt(cmdArr.commandParam)

                        };

                        self._timerChangeList.push(new_Obj);


                        break;


                    default:

                        if(genConfig.DebugMessages == true && genConfig.DebugMessagesSimulation == true) {

                            console.log("# xbee-serialsim # Unknown ");
                        }


                        break;


                }




            }




            break;

        default:

            console.log('UNKNOWN_PACKET_TYPE');

            break;
    }


};




exports.NodeSim = NodeSim;


